applications/tari_console_wallet/src/grpc/wallet_grpc_server.rs

```diff
@@ -0,0 +1,66 @@
+use std::sync::Arc;
+use tari_app_grpc::tari_rpc::{wallet_server, GetCoinbaseRequest, GetCoinbaseResponse};
+use tari_core::transactions::transaction::Transaction;
+use tari_wallet::{
+    contacts_service::storage::sqlite_db::ContactsServiceSqliteDatabase,
+    output_manager_service::storage::sqlite_db::OutputManagerSqliteDatabase,
+    storage::sqlite_db::WalletSqliteDatabase,
+    transaction_service::storage::sqlite_db::TransactionServiceSqliteDatabase,
+    Wallet,
+};
+use tokio::sync::RwLock;
+use tonic::{Request, Response, Status};
+
+pub struct WalletGrpcServer {
+    wallet: Arc<
+        RwLock<
+            Wallet<
+                WalletSqliteDatabase,
+                TransactionServiceSqliteDatabase,
+                OutputManagerSqliteDatabase,
+                ContactsServiceSqliteDatabase,
+            >,
+        >,
+    >,
+}
+
+impl WalletGrpcServer {
+    pub fn new(
+        wallet: Arc<
+            RwLock<
+                Wallet<
+                    WalletSqliteDatabase,
+                    TransactionServiceSqliteDatabase,
+                    OutputManagerSqliteDatabase,
+                    ContactsServiceSqliteDatabase,
+                >,
+            >,
+        >,
+    ) -> Self
+    {
+        Self { wallet }
+    }
```

Comment:
You'll be able to avoid the lock entirely by cloning the `transaction_service` each time you need to use it. Essentially you're cloning an `mpsc::UnboundedSender` which is very cheap. 

```suggestion
    pub fn new(
        wallet: Arc<
                Wallet<
                    WalletSqliteDatabase,
                    TransactionServiceSqliteDatabase,
                    OutputManagerSqliteDatabase,
                    ContactsServiceSqliteDatabase,
                >,
        >,
    ) -> Self
    {
        Self { wallet }
    }
```

Depending on how `TransactionService` is implemented, you should also be able to get some concurrency gains.

Reply:
Perhaps worth looking into this at a later stage. I don't disagree that it may be more optimal, but there is still quite a lot to implement, and at this stage it's a fairly heavy lock. Conceptually I was thinking: The GRPC server and TUI will be using the wallet. Some parts of the GRPC are only going to need read access to it, some need write. In the TUI, the display cycle only reads from the wallet, and when you press a button, or invoke an action you need write access to it. You may be able to go into finer grain locking, or use a non-locking strategy, but as I said, not all of the functions are implemented yet, so you may find that you constantly changing these locks.

Reply:
As discussed this architectural comment will be addressed in subsequent PRs